### DISCUSSION TITLE
[Art competition] Starting program analysis

### DISCUSSION THREAD

**Discussion by zzz845zz** [prosyslab-classroom/cs524-program-analysis#276] [Art competition]
![DALL·E 2022-11-29 23 48 41 - A group of warriors who go to challenge the unknown Devil Castle with the given information](https://user-images.githubusercontent.com/16317528/204576920-789e14f1-b7d6-4895-b645-a5548e7307fd.png)

## Short Description

For me, [*Starting program analysis*] means [*A group of warriors goes to challenge the demon king's castle with the given information*]".

## Long Description (up to 3 sentences)

The goal of program analysis is to find out whether the desired or undesired behaviors occur in a program. However, you may have to trial and error many times because the provided information by the client may not be sufficient to analyze the program. Will you be able to refine the given information and successfully defeat the demon king called program analysis?

**Comment by KihongHeo** [comment]
> the provided information by the client may not be sufficient to analyze the program

What do you mean?
What do you mean by "refine the given information"? Any example?

**Comment by zzz845zz** [comment]
There is no concrete example, but I imagined a real-world scenario like this:

- Clients: requests the analysis with some helpful(?) information.
  - "Hey analysts, please analyze (or make an analyzer for) my program to make sure it works well."
  - "Maybe the sign domain is enough to abstract our program because we don't use `add(+)` and `sub(-)` operations."
- Analysts: refine the given information through trial and error.
  - After try 1: "Sign domain is not enough. We need interval domain."
  - -> After try 2: "Interval domain is not enough, too. Let's give a threshold for widening."
  - -> After try 3: "Interval domain with threshold is not enough, too. We need another way to make it more precise."
  - -> ...
  - -> After try n: “Finally..!”

**Comment by KihongHeo** [comment]
Oh. Good. Have you heard about "abstraction refinement"? That is exactly the same process as you mentioned. It can be done automatically but may not terminate. Take a look at that if you are interested!
